src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,140 @@
+            throw new ArrayListIndexOutOfBoundsException(BUG_INDEX);
+        }
+        if (elementData.length == size) {
+            resize(lengthZoom());
+        }
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        int numNew = list.size();
+        if (numNew == 0) {
+            return;
+        }
+        if (numNew > (elementData.length - size)) {
+            resize((size + numNew));
+        }
+        for (int i = 0; i < numNew; i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(BUG_INDEX);
+        }
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(BUG_INDEX);
+        }
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(BUG_INDEX);
+        }
+        final Object[] es = elementData;
+        T oldValue = (T) es[index];
+        fastRemove(es, index);
+        return oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        final Object[] es = elementData;
+        int index = -1;
+        if (element == null) {
+            for (int i = 0; i < size; i++) {
+                if (es[i] == null) {
+                    index = i;
+                    break;
+                }
+            }
+        } else {
+            for (int i = 0; i < size; i++) {
+                if (element.equals(es[i])) {
+                    index = i;
+                    break;
+                }
+            }
+        }
+        try {
+            T result = (T) es[index];
+        } catch (IndexOutOfBoundsException e) {
+            throw new NoSuchElementException(BUG_FOUND);
+        }
+        T oldValue = (T) es[index];
+        fastRemove(es, index);
+        return oldValue;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    @Override
+    public String toString() {
+        return "ArrayList{" + "elementData=" + Arrays.toString(elementData)
+                + ", size=" + size + '}';
+    }
+
+    public int lengthZoom() {
+        return (int) (elementData.length * ZOOM_SIZE);
+    }
+
+    public void resize(int newLength) {
+        Object[] newArray = new Object[newLength];
+        System.arraycopy(elementData, 0, newArray, 0, size);
+        elementData = newArray;
+    }
```

Comment:
The resize method should always increase the size of the array by the same value